sbin/mount_vnd/mount_vnd.8
.\"	$OpenBSD: mount_vnd.8,v 1.21 2014/05/30 16:53:02 tedu Exp $
.\"
.\" Copyright (c) 1993 University of Utah.
.\" Copyright (c) 1980, 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Systems Programming Group of the University of Utah Computer
.\" Science Department.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)vnconfig.8	8.1 (Berkeley) 6/5/93
.\"
.\"
.\" Copyright (c) 2007 Alexander von Gernler <grunk@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: May 30 2014 $
.Dt MOUNT_VND 8
.Os
.Sh NAME
.Nm mount_vnd ,
.Nm vnconfig
.Nd configure vnode disks
.Sh SYNOPSIS
.Nm mount_vnd
.Bk -words
.Op Fl o Ar options
.Op Fl t Ar disktype
.Ar image
.Ar vnd_dev
.Ek
.Nm vnconfig
.Bk -words
.Op Fl cluv
.Op Fl t Ar disktype
.Ar vnd_dev
.Ar image
.Ek
.Sh DESCRIPTION
The
.Nm vnconfig
command configures vnode pseudo disk devices.
It will associate (or disassociate) the special file
.Ar vnd_dev
with the regular file
.Ar image ,
allowing the latter to be accessed as though it were a disk.
.Pp
.Nm mount_vnd
works similarly to
.Nm vnconfig ,
but it provides an interface that can be used by the
.Xr fstab 5
infrastructure, so that an
.Ar image
file can be configured to a device
.Ar node
while booting.
.Nm mount_vnd
defaults to the
.Pa vnd0
device.
.Pp
For
.Xr fstab 5
lines with type
.Dq ffs ,
the
.Dq noauto
option must be set to prevent a
.Xr mount 8
of the FFS partitions
before the necessary vnd devices are configured.
Also, the
.Dq fs_passno
field has to be set to 0 to prevent
.Xr fsck 8
from checking the file system for the same reasons.
.Pp
.Nm mount_vnd
is invoked by
.Xr mount 8
when using the following syntax:
.Bd -ragged -offset 4n
.Nm mount Op options
-t vnd
.Ar image Ar node
.Ed
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c
.Nm vnconfig
only.
Configures the device.
If successful, references to
.Ar vnd_dev
will access the contents of
.Ar image .
This is the default operation.
.It Fl l
.Nm vnconfig
only.
List the vnd devices and indicate which ones are in use.
If a specific
.Ar vnd_dev
is given, then only that one will be described.
.It Fl o Ar options
.Nm mount_vnd
only.
Options are specified with a
.Fl o
flag followed by a comma separated string of options.
See the
.Xr mount 8
man page for possible options and their meanings.
.Pp
At the moment,
.Fl o
is only here for compatibility reasons, but no use is made of supplied
options.
.It Fl t Ar disktype
Specify a
.Ar disktype
entry from the
.Xr disktab 5
database.
The
.Ar vnd_dev
will have the sector size, sectors per track, and tracks per cylinder values
of the specified
.Ar disktype .
The defaults are 512-byte sectors, 100 sectors per track and 1 track per
cylinder.
.It Fl u
.Nm vnconfig
only.
Unconfigures a
.Ar vnd_dev .
.It Fl v
.Nm vnconfig
only.
Print messages to stdout describing actions taken.
.El
.Sh FILES
.Bl -tag -width /etc/rvnd?? -compact
.It Pa /dev/{,r}vnd*
.El
.Sh EXAMPLES
Configure a CD-ROM or DVD image file as vnode disk vnd0
and mount the ISO 9660 file system contained in it:
.Bd -literal -offset indent
# vnconfig vnd0 /tmp/diskimage
# mount -t cd9660 /dev/vnd0c /mnt
.Ed
.Pp
An equivalent
.Xr fstab 5
entry is:
.Bd -literal -offset indent
/tmp/diskimage /dev/vnd0c vnd    rw,noauto 0 0
/dev/vnd0a     /mnt       cd9660 rw,noauto 0 0
.Ed
.Pp
Unmount the file system and unconfigure the vnode device:
.Bd -literal -offset indent
# umount /mnt
# vnconfig -u vnd0
.Ed
.Pp
A sample of commands to put in
.Xr rc.local 8
in order to have vnd images configured at boot time is listed here:
.Bd -literal -offset indent
# Example for automatically configuring a vnd device on startup
echo "configuring vnd devices:"
mount /dev/vnd0c
fsck -p /dev/rvnd0a
mount /mnt
.Ed
.Pp
Mounting images during the first pass of
.Xr fsck 8
and
.Xr mount 8
is not possible, because the image to be configured to a vnd itself
resides on a file system that first has to be checked and mounted.
.Sh SEE ALSO
.Xr vnd 4 ,
.Xr disktab 5 ,
.Xr fstab 5 ,
.Xr mount 8 ,
.Xr swapon 8 ,
.Xr umount 8
.Sh HISTORY
The
.Nm
command first appeared in
.Ox 4.2 .
.Sh CAVEATS
The
.Sq c
partition of a vnd image should not be used.
When a superblock becomes damaged,
.Xr fsck_ffs 8
needs information contained in the disklabel
to determine the location of alternate superblocks.
This information is not available when directly using the
.Sq c
partition,
so checking the file system image will fail.
.Pp
Swapping to vnd devices is no longer supported.


sbin/mount_vnd/mount_vnd.c
/*	$OpenBSD: mount_vnd.c,v 1.16 2014/10/29 21:30:10 tedu Exp $	*/
/*
 * Copyright (c) 1993 University of Utah.
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * the Systems Programming Group of the University of Utah Computer
 * Science Department.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * from: Utah $Hdr: vnconfig.c 1.1 93/12/15$
 *
 *	@(#)vnconfig.c	8.1 (Berkeley) 12/15/93
 */

#include <sys/param.h>
#include <sys/ioctl.h>
#include <sys/mount.h>
#include <sys/stat.h>
#include <sys/disklabel.h>

#include <dev/biovar.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <util.h>

#define DEFAULT_VND	"vnd0"

#define VND_CONFIG	1
#define VND_UNCONFIG	2
#define VND_GET		3

int verbose = 0;
int run_mount_vnd = 0;

__dead void	 usage(void);
int		 config(char *, char *, int, struct disklabel *);
int		 getinfo(const char *);
int		 getinfo_all(void);

int
main(int argc, char **argv)
{
	int	 ch, rv, action, opt_c, opt_l, opt_u;
	char	*mntopts;
	extern char *__progname;
	struct disklabel *dp = NULL;

	if (strcasecmp(__progname, "mount_vnd") == 0)
		run_mount_vnd = 1;

	opt_c = opt_l = opt_u = 0;
	mntopts = NULL;
	action = VND_CONFIG;

	while ((ch = getopt(argc, argv, "clo:t:uv")) != -1) {
		switch (ch) {
		case 'c':
			opt_c = 1;
			break;
		case 'l':
			opt_l = 1;
			break;
		case 'o':
			mntopts = optarg;
			break;
		case 't':
			dp = getdiskbyname(optarg);
			if (dp == NULL)
				errx(1, "unknown disk type: %s", optarg);
			break;
		case 'u':
			opt_u = 1;
			break;
		case 'v':
			verbose = 1;
			break;
		default:
			usage();
			/* NOTREACHED */
		}
	}
	argc -= optind;
	argv += optind;

	if (opt_c + opt_l + opt_u > 1)
		errx(1, "-c, -l and -u are mutually exclusive options");

	if (opt_l)
		action = VND_GET;
	else if (opt_u)
		action = VND_UNCONFIG;
	else
		action = VND_CONFIG;	/* default behavior */

	if (action == VND_CONFIG && argc == 2) {
		int ind_raw, ind_reg;

		/* fix order of arguments. */
		if (run_mount_vnd) {
			ind_raw = 1;
			ind_reg = 0;
		} else {
			ind_raw = 0;
			ind_reg = 1;
		}
		rv = config(argv[ind_raw], argv[ind_reg], action, dp);
	} else if (action == VND_UNCONFIG && argc == 1)
		rv = config(argv[0], NULL, action, NULL);
	else if (action == VND_GET && argc == 1) {
		rv = getinfo(argv[0]);
		if (rv == -1)
			err(4, "BIOCLOCATE");
	} else if (action == VND_GET && argc == 0)
		rv = getinfo_all();
	else
		usage();

	exit(rv);
}

int
getinfo(const char *vname)
{
	struct bio_locate bl;
	struct bioc_vndget vget;
	int fd;

	if ((fd = open("/dev/bio", O_RDWR)) < 0)
		err(1, "/dev/bio");

	bl.bl_name = (char *)vname;
	if (ioctl(fd, BIOCLOCATE, &bl) < 0) {
		if (errno == ENOENT)
			return (-1);
		err(4, "BIOCLOCATE");
	}

	vget.bvg_bio.bio_cookie = bl.bl_bio.bio_cookie;

	if (ioctl(fd, BIOCVNDGET, &vget) == -1)
		err(1, "ioctl: %s", vname);

	fprintf(stdout, "%s: ", vname);

	if (!vget.bvg_ino)
		fprintf(stdout, "not in use\n");
	else
		fprintf(stdout, "covering %s on %s, inode %llu\n",
		    vget.bvg_file, devname(vget.bvg_dev, S_IFBLK),
		    (unsigned long long)vget.bvg_ino);

	close(fd);
	return (0);
}

int
getinfo_all(void)
{
	char vname[7];
	unsigned char i;
	int rv;

	/*
	 * We don't know how many vnd devices are configured in the
	 * kernel, so iterate until we hit a non-existent.
	 */
	for (i = 0; i < UCHAR_MAX; i++) {
		(void)snprintf(vname, sizeof(vname), "vnd%d", i);
		rv = getinfo(vname);
		if (rv != 0) {
			if (rv == -1)	/* no such file */
				return (0);
			return (rv);
		}
	}

	return (0);
}

int
config(char *dev, char *file, int action, struct disklabel *dp)
{
	struct bio_locate bl;
	struct bioc_vndset vset;
	struct bioc_vndclr vclr;
	int fd, rv = -1;

	if ((fd = open("/dev/bio", O_RDWR)) < 0)
		err(4, "/dev/bio");

	bl.bl_name = dev;
	if (ioctl(fd, BIOCLOCATE, &bl) < 0)
		err(4, "BIOCLOCATE");

	/*
	 * Clear (un-configure) the device
	 */
	if (action == VND_UNCONFIG) {
		vclr.bvc_bio.bio_cookie = bl.bl_bio.bio_cookie;

		rv = ioctl(fd, BIOCVNDCLR, &vclr);
		if (rv)
			warn("BIOCVNDCLR");
		else if (verbose)
			printf("%s: cleared\n", dev);
	}
	/*
	 * Configure the device
	 */
	if (action == VND_CONFIG) {
		vset.bvs_bio.bio_cookie = bl.bl_bio.bio_cookie;
		vset.bvs_file = file;
		vset.bvs_secsize =
		    (dp && dp->d_secsize) ? dp->d_secsize : DEV_BSIZE;
		vset.bvs_nsectors =
		    (dp && dp->d_nsectors) ? dp->d_nsectors : 100;
		vset.bvs_ntracks = (dp && dp->d_ntracks) ? dp->d_ntracks : 1;

		rv = ioctl(fd, BIOCVNDSET, &vset);
		if (rv)
			warn("BIOCVNDSET");
		else if (verbose)
			printf("%s: %llu bytes on %s\n", dev, vset.bvs_size,
			    file);
	}

	close(fd);
	fflush(stdout);
	return (rv < 0);
}

__dead void
usage(void)
{
	extern char *__progname;

	if (run_mount_vnd)
		(void)fprintf(stderr,
		    "usage: %s [-o options] [-t disktype] image vnd_dev\n",
		    __progname);
	else
		(void)fprintf(stderr,
		    "usage: %s [-cluv] [-t disktype] vnd_dev image\n",
		    __progname);

	exit(1);
}
